lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -838,36 +838,67 @@ class ADContext {
   SILPassManager &getPassManager() const { return passManager; }
   Lowering::TypeConverter &getTypeConverter() { return module.Types; }
 
-  SmallVectorImpl<DifferentiableFunctionInst *> &
-  getDifferentiableFunctionInsts() {
-    return differentiableFunctionInsts;
+  /// Returns true if there are `DifferentiableFunctionInst` to process.
+  bool hasDifferentiableFunctionInsts() const {
+    return !differentiableFunctionInsts.empty();
   }
 
-  SmallPtrSetImpl<DifferentiableFunctionInst *> &
-  getProcessedDifferentiableFunctionInsts() {
-    return processedDifferentiableFunctionInsts;
+  /// Pick and remove a `DifferentiableFunctionInst` for processing.
+  /// Returns nullptr if none are available for processing.
+  DifferentiableFunctionInst* pickAndRemoveDifferentiableFunctionInst() {
+    if (differentiableFunctionInsts.empty()) return nullptr;
+    auto *dfi = differentiableFunctionInsts.back();
+    differentiableFunctionInsts.pop_back();
+    return dfi;
+  }
+
+  void addDifferentiableFunctionInst(DifferentiableFunctionInst* dfi) {
+    differentiableFunctionInsts.push_back(dfi);
+  }
+
+  bool
+  isDifferentiableFunctionInstProcessed(DifferentiableFunctionInst *dfi) const {
+    return processedDifferentiableFunctionInsts.count(dfi) > 0;
+  }
+
+  void
+  markDifferentiableFunctionInstAsProcessed(DifferentiableFunctionInst *dfi) {
+    processedDifferentiableFunctionInsts.insert(dfi);
   }
 
-  llvm::SmallMapVector<SILDifferentiabilityWitness *, DifferentiationInvoker,
-                       32> &
-  getInvokers() {
+  const llvm::SmallMapVector<SILDifferentiabilityWitness *,
+                             DifferentiationInvoker, 32> &
+  getInvokers() const {
     return invokers;
   }
 
-  DenseMap<DifferentiableFunctionInst *, unsigned> &getResultIndices() {
-    return resultIndices;
+  void addInvoker(SILDifferentiabilityWitness *witness) {
+    assert(!invokers.count(witness) &&
+           "Differentiability witness already has an invoker");
+    invokers.insert({witness, DifferentiationInvoker(witness)});
+  }
+
+  /// Returns the index for `dfi` if found in this context. Otherwise,
+  /// sets the index to zero and returns it.
+  unsigned getResultIndex(DifferentiableFunctionInst* dfi) {
+    return resultIndices[dfi];
+  }
+
+  /// Updates the index for `dfi` in this context.
+  unsigned updateResultIndex(DifferentiableFunctionInst* dfi, unsigned index) {
+    return resultIndices[dfi] = index;
   }
 
   DenseMap<ApplyInst *, NestedApplyInfo> &getNestedApplyInfo() {
     return nestedApplyInfo;
   }
 
-  SmallVector<SILFunction *, 32> &getGeneratedFunctions() {
-    return generatedFunctions;
+  void addGeneratedFunction(SILFunction* function) {
```

Comment:
I think "record" is slightly more apt.
```suggestion
  void recordGeneratedFunction(SILFunction* function) {
```